File: tests/integration/publish/peer.js
'use strict';

// Peer clients to the publish integration tests.
// Usage: node peer.js <test name> <router url> <test realm>

const autobahn = require('autobahn');

function publishOneMessageTestPeer(routerUrl, testRealm) {
    const connection = new autobahn.Connection({'url': routerUrl, 'realm': testRealm});
    connection.onopen = (session) => {
        session.subscribe('org.test.topic1', (args) => {
            // This test expects to receive one publication to this channel.
            if (args[0] === 42) {
                console.log('test passed');
            } else {
                console.log('test failed');
            }
        });

        // We've subscribed to all channels; tell the test to proceed.
        console.log('ready');
    };
    connection.open();
    console.error("opening connection");
}

function publishMultipleMessagesTestPeer(routerUrl, testRealm) {
    const connection = new autobahn.Connection({'url': routerUrl, 'realm': testRealm});
    connection.onopen = (session) => {
        let recvTopic1 = false;
        let recvTopic2Msg1 = false;
        let recvTopic2Msg2 = false;

        session.subscribe('org.test.topic1', (args, kwargs) => {

        });

        session.subscribe('org.test.topic2', (args, kwargs) => {

        });

        // We've subscribed to all channels; tell the test to proceed.
        console.log('ready');
    };
    connection.open();
}


const testName = process.argv[2];
const routerUrl = process.argv[3];
const testRealm = process.argv[4];
console.error("starting peer:", testName, routerUrl, testRealm);

switch (testName) {
    case "publishOneMessage": {
        publishOneMessageTestPeer(routerUrl, testRealm);
        break;
    }
    case "publishMultipleMessages": {
        publishMultipleMessagesTestPeer(routerUrl, testRealm);
        break;
    }
}

File: tests/integration/publish/peer.py
#!/usr/bin/env python3

import sys
from autobahn.asyncio.component import Component, run


async def publish_one_message_on_joined(session, details):
    # We expect another client to publish a single message to 'org.test.topic1'.
    def received_msg(event):
        print(event)
        print('test passed: received message')

    try:
        await session.subscribe(received_msg, 'org.test.topic1')
        print('ready: subscribed to org.test.topic1')

    except Exception as e:
        print('test failed: unable to subscribe to topic:', e)


def run_test_peer(component, test_name):
    print('running test peer:', test_name)

    # pylint: disable=unused-variable
    @component.on_join
    async def joined(session, details):
        print('test component joined:', test_name, session, details)

        if test_name == 'publishOneMessage':
            await publish_one_message_on_joined(session, details)

        else:
            print('test failed: unknown test name')

    run([component])


if __name__ == '__main__':
    test_name = sys.argv[1]
    router_url = sys.argv[2]
    test_realm = sys.argv[3]

    component = Component(
        transports=router_url,
        realm=test_realm,
    )
    run_test_peer(component, test_name)
